libz-rs-sys/src/gz.rs

```diff
@@ -249,28 +289,155 @@ unsafe fn gzopen_help(path: *const c_char, fd: c_int, mode: *const c_char) -> gz
+unsafe fn gz_error(state: &mut GzState, err: c_int, msg: *const c_char) {
+    if !state.msg.is_null() {
+        // NOTE: zlib-ng has a special case here: it skips the deallocation if
+        // state.err == Z_MEM_ERROR. However, we always set state.msg to null
+        // when state.err is set to Z_MEM_ERROR, so that case is unreachable
+        // here.
+        unsafe { deallocate_cstr(state.msg.cast_mut()) };
+        state.msg = ptr::null_mut();
+    }
+
+    // On error, set state.have to 0 so that the `gzgetc()` C macro fails
+    if err != Z_OK && err != Z_BUF_ERROR {
+        state.have = 0;
+    }
+
+    // Set the error code, and if no message, then done
+    state.err = err;
+    if msg.is_null() {
+        return;
+    }
+
+    // For an out of memory error, don't bother trying to allocate space for an error string.
+    // ([`gzerror`] will provide literal string as a special case for OOM errors.)
+    if err == Z_MEM_ERROR {
+        return;
+    }
+
+    // Format the error string to include the file path.
+    // Safety: `gzopen` and `gzdopen` ensure that `state.path` is a non-null C string,
+    //          the caller of this function is reponsible for ensuring that `msg` is a C string,
+    //          and we exit this function above if `msg` is null.
+    let sep = b": \0".as_ptr().cast::<c_char>();
+    let buf = &mut [0u8; 17];
+    state.msg = match state.source {
+        Source::Path(path) => unsafe { gz_strcat(&[path, sep, msg]) },
+        Source::Fd(fd) => unsafe { gz_strcat(&[fd_path(buf, fd).as_ptr(), sep, msg]) },
+    };
+
+    if state.msg.is_null() {
+        state.err = Z_MEM_ERROR;
+    }
+}
+
 // Deallocate a GzState structure and all heap-allocated fields inside it.
 //
 // # Safety
 //
 // - The `state` object and all heap-allocated fields within it must have been obtained
 //   using `ALLOCATOR`.
 // - The caller must not use the `state` after passing it to this function.
-unsafe fn free_state(state: &mut GzState) {
+unsafe fn free_state(state: *mut GzState) {
+    // Safety: `deallocate_cstr` accepts null pointers or C strings, and in this
+    // module we use only `ALLOCATOR` to allocate strings assigned to these fields.
     unsafe {
-        if !state.path.is_null() {
-            ALLOCATOR.deallocate::<c_char>(state.path.cast_mut(), libc::strlen(state.path) + 1);
+        match (*state).source {
+            Source::Path(path) => deallocate_cstr(path.cast_mut()),
+            Source::Fd(_) => { /* fd is owned by the caller */ }
         }
+        deallocate_cstr((*state).msg.cast_mut());
+    }
+    // Safety: The caller has ensured that `state` was allocated using `ALLOCATOR`.
+    unsafe {
         ALLOCATOR.deallocate::<GzState>(state, 1);
     }
 }
 
+// Free a string that was allocated with `ALLOCATOR`
+//
+// # Safety
+//
+// `s` must be either null or a null-terminated C string that was allocated with `ALLOCATOR`.
```

Comment:
the length of the string (plus the null terminator) must also match the length of the allocation